presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java

```diff
@@ -13,202 +13,65 @@
             return false;
         }
-
-        if (leftPositions == null || leftPositions.length < leftPositionCount) {
-            leftPositions = new int[leftPositionCount * 2];
-        }
-
-        if (rightPositions == null || rightPositions.length < rightPositionCount) {
-            rightPositions = new int[rightPositionCount * 2];
-        }
-
-        for (int i = 0; i < leftPositionCount; i++) {
-            leftPositions[i] = i;
-        }
-        for (int i = 0; i < rightPositionCount; i++) {
-            rightPositions[i] = i;
+        Block lookArray = leftArray;
+        Block itrArray = rightArray;
+        if (leftPositionCount > rightPositionCount) {
+            lookArray = rightArray;
+            itrArray = leftArray;
         }
-        IntArrays.quickSort(leftPositions, 0, leftPositionCount, intBlockCompare(type, leftArray));
-        IntArrays.quickSort(rightPositions, 0, rightPositionCount, intBlockCompare(type, rightArray));
 
-        int leftCurrentPosition = 0;
-        int rightCurrentPosition = 0;
-        while (leftCurrentPosition < leftPositionCount && rightCurrentPosition < rightPositionCount) {
-            if (leftArray.isNull(leftPositions[leftCurrentPosition]) || rightArray.isNull(rightPositions[rightCurrentPosition])) {
-                // Nulls are in the end of the array. Non-null elements do not overlap.
-                return null;
+        boolean hasNull = false;
+        TypedSet typedSet = new TypedSet(type, lookArray.getPositionCount(), "arraysOverlap");
+        for (int i = 0; i < lookArray.getPositionCount(); i++) {
+            if (lookArray.isNull(i)) {
+                hasNull = true;
+                continue;
             }
-
-            int compareValue;
-            try {
-                compareValue = type.compareTo(leftArray, leftPositions[leftCurrentPosition], rightArray, rightPositions[rightCurrentPosition]);
-            }
-            catch (NotSupportedException e) {
-                throw new PrestoException(NOT_SUPPORTED, e.getMessage(), e);
-            }
-
-            if (compareValue > 0) {
-                rightCurrentPosition++;
-            }
-            else if (compareValue < 0) {
-                leftCurrentPosition++;
-            }
-            else {
-                return true;
-            }
-        }
-        return leftArray.isNull(leftPositions[leftPositionCount - 1]) || rightArray.isNull(rightPositions[rightPositionCount - 1]) ? null : false;
-    }
-
-    public Boolean genericArraysOverlap(Block leftArray, Block rightArray, AbstractType type)
-    {
-        int leftSize = leftArray.getPositionCount();
-        int rightSize = rightArray.getPositionCount();
-
-        if (leftSize == 0 || rightSize == 0) {
-            return false;
-        }
-
-        if (leftLongArray == null || leftLongArray.length < leftSize) {
-            leftLongArray = new long[leftSize * 2];
+            typedSet.add(lookArray, i);
```

Comment:
Or you can do addNonNull here and keep the nullcheck

Reply:
Rewrote the logic by adding helper method in TypedSet to check if it containsNullElement().